### PR TITLE
Treat `--upgrade-packages` PKGSPECs as constraints (not just minimums), consistently

### LINKED ISSUE #1550: --upgrade-packages fails to constrain subdeps, if either absent from a preexisting output file, or if --upgrade is also passed
The documentation states that the `--upgrade` and `--upgrade-packages` [can be used together](https://github.com/jazzband/pip-tools/blob/6.4.0/README.rst#updating-requirements), however the `--upgrade-packages` flag is ignored when `--upgrade` is supplied.

#### Environment Versions

1. OS Type: macOS and Ubuntu 18.04
2. Python version: `python 3.9.0`
3. pip version: `pip 21.3.1`
4. pip-tools version: `pip-compile 6.4.0`

#### Steps to replicate

_Using different args to the documentation, as `requests` is not in the `django` requirements, and `requests<=3.0.0` does not affect the "latest" release._

```bash
# Create requirements.in file
echo "django" > requirements.in
# Run pip-compile
pip-compile --upgrade --upgrade-package 'sqlparse<=0.4.0'
```

#### Expected result

```
asgiref==3.4.1
    # via django
django==4.0.1
    # via -r requirements.in
sqlparse==0.4.0
    # via django
```

#### Actual result

```
asgiref==3.4.1
    # via django
django==4.0.1
    # via -r requirements.in
sqlparse==0.4.2 
    # via django
```

`sqlparse` is not restricted to `<=0.4.0`. If I omit the `--upgrade` flag, only supplying `--upgrade-packages 'sqlparse<=0.4.0'`, I get the expected result (but my other packages are not upgraded to later versions). This also doesn't work if the argument is `sqlparse==0.4.0`.

My current workaround is just to delete the requirements file(s) before running `pip-compile`, and omitting the `--upgrade` flag so the existing files aren't used as a cache. I suspect either the `--upgrade-packages` constraint logic is inside the `if not upgrade...` [conditional](https://github.com/jazzband/pip-tools/blob/6.4.0/piptools/scripts/compile.py#L354), or that `--upgrade-packages` only [extend](https://github.com/jazzband/pip-tools/blob/6.4.0/piptools/scripts/compile.py#L437), not constrain the complete set of packages.

This feature should be permitted, or the documentation changed to avoid stating that this is possible

### PR DESCRIPTION
**Pull request by AndydeCleyre**
Fixes #1550

This aims to address two problems:

- When `--upgrade` and `--upgrade-packages PKGSPEC` are used together, and `PKGSPEC`'s package is not in the input file (but is a subdependency), then that package is upgraded beyond `PKGSPEC`
- When `--upgrade-packages PKGSPEC` is used, and `PKGSPEC`'s package is in neither the input file nor a preexisting output file (but is a subdependency), then that package is upgraded beyond `PKGSPEC`

In other words:

> `--upgrade-packages` fails to constrain subdeps, if either absent from a preexisting output file, or if `--upgrade` is also passed

Some of the current behavior was informed by discussion at #759.

However my understanding now is that it's useful, expected, and without significant cost to additionally treat `--upgrade-packages` specs as constraints, thereby resolving the above cases.

##### Contributor checklist

- [ ] Provided the tests for the changes.
- [ ] Assure PR title is short, clear, and good to be included in the user-oriented changelog

##### Maintainer checklist

- [ ] Assure one of these labels is present: `backwards incompatible`, `feature`, `enhancement`, `deprecation`, `bug`, `dependency`, `docs` or `skip-changelog` as they determine changelog listing.
- [ ] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by AndydeCleyre** [comment]
This currently has at least one bug:

`requirements.in`:
```ini
ptrender
```

`requirements.txt`:
```ini
sqlparse==0.4.0
```

```console
$ pip-compile --upgrade-package 'sqlparse<=0.4.0' requirements.in
```

`requirements.txt`:
```ini
plumbum==1.7.2
    # via ptrender
ptrender==0.0.3
    # via -r requirements.in
pyratemp==0.3.2
    # via ptrender
sqlparse==0.4.0
```

**Comment by AndydeCleyre** [comment]
Thank you @jammie19 !

I'm refining my understanding of what we're solving here, and have tried to articulate it as two problem cases, bulleted above.

In terms of tests, I think the second bullet is now covered by a simple change to the existing `test_upgrade_packages_option_subdependency`.

I'm going to have a look and see if a similar tiny test adjustment will be appropriate for the first bullet as well.

**Comment by AndydeCleyre** [comment]
OK I think the new parametrization for `test_upgrade_packages_version_option_and_upgrade` covers the first bullet, so I'm going to promote this out of draft state.

**Review by richafrank — COMMENTED** [review]
Thanks for working to restore this behavior @AndydeCleyre !

File: piptools/scripts/compile.py [hunk 1]
```diff
@@ -433,9 +450,8 @@ def cli(
         key_from_ireq(ireq) for ireq in constraints if not ireq.constraint
     }
 
-    allowed_upgrades = primary_packages | existing_pins_to_upgrade
```
[the remark this answers is not in the document]
Comment (AndydeCleyre) [reply piptools/scripts/compile.py:436]: Excellent, thanks! Removed in 25a4e15cef7c4e39d6bfcb9625d19ea45dbf23f6

File: piptools/scripts/compile.py [hunk 2]
```diff
@@ -423,6 +423,23 @@ def cli(
                 )
             )
 
+    if upgrade_packages:
+        constraints_file = tempfile.NamedTemporaryFile(mode="wt", delete=False)
```
[the remark this answers is not in the document]
Comment (AndydeCleyre) [reply piptools/scripts/compile.py:422]: Deleting would be ideal, and I originally had it getting deleted, but it always failed the Windows CI, somehow getting removed before some component tried to access it. 

Thanks! Let's try this and see.

[the remark this answers is not in the document]
Comment (atugushev) [reply piptools/scripts/compile.py:422]: > but it always failed the Windows CI, somehow getting removed before some component tried to access it.

This is known issue with Windows. See also https://github.com/jazzband/pip-tools/pull/723. So `delete=False` and deleting the file later should be fine.

[the remark this answers is not in the document]
Comment (AndydeCleyre) [reply piptools/scripts/compile.py:422]: Yeah, the change breaks Windows. I 've reverted for now, but the test logs are available and I think indicate certain compile outputs being empty.

[the remark this answers is not in the document]
Comment (AndydeCleyre) [reply piptools/scripts/compile.py:422]: Excellent, thanks @atugushev

**Review by atugushev — APPROVED** [review]
Finally got to this PR. That looks awesome, thanks!

**Pull request merged**